### ISSUE TITLE
Collection expressions: Report warning when the compiler generates a heap allocation for a ref struct

### ISSUE THREAD

**Issue by cston** [dotnet/roslyn#69693]
Report a warning when the compiler generates a heap allocation for a collection expression with a `ref struct` target type.

For instance, the collection expression in the following will result in a heap allocation for an `object[]` when compiling with .NET 7 or earlier.
```csharp
using System;

class Program
{
    static void Main()
    {
        Print([1, 2, 3]);
    }
    
    static void Print(ReadOnlySpan<object> items)
    {
    }
}
```

See https://github.com/dotnet/roslyn/pull/69412#discussion_r1303396806

**Comment by RikkiGibson** [comment]
We think this warning should be disabled by default, much like the "implicit field assignment" warnings we did for the struct auto-default feature.
